File: AlhambraXR/Assets/Scripts/Data/Annotation/Annotation.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using UnityEngine;

[Serializable]
public class Annotation
{
    // pure data
    [SerializeField]
    public AnnotationInfo info = null;
    [SerializeField]
    public AnnotationRenderInfo renderInfo = new AnnotationRenderInfo();

    [NonSerialized]
    internal List<AnnotationJoint> joints = new List<AnnotationJoint>();

    // if the annotation is stored in both sides, then we do not sync them
    [NonSerialized]
    internal bool isLocalData = false;

    // control data
    public IReadOnlyList<AnnotationJoint> Joints => joints;

    [SerializeField]
    private AnnotationID id;

    public AnnotationID ID
    {
        get=>id;
    }

    public Annotation(AnnotationID _id)
    {
        id = _id;
    }

    public void RemoveJoint(AnnotationJoint joint)
    {
        joint.RemoveAnnotation(this);
    }

    public void AddJoint(AnnotationJoint joint)
    {
        joint.AddAnnotation(this);
    }

    public AnnotationJoint FindJoint(int jointID)
    {
        return joints.Find(x => x.ID == jointID);
    }

    public bool HasJoint(int jointID)
    {
        return FindJoint(jointID) != null;
    }

    public bool IsValid
    {
        get => info != null && ID.IsValid;
    }

    public void PostDeserialize()
    {
        if(joints==null)
        {
            joints = new List<AnnotationJoint>();
        }
    }




}

File: Alhambra_Android/app/src/main/java/com/alhambra/dataset/data/Annotation.java
package com.alhambra.dataset.data;

import androidx.annotation.Nullable;

import java.util.Collections;
import java.util.HashSet;
import java.util.Set;

public class Annotation {
    public AnnotationInfo info;
    public AnnotationRenderInfo renderInfo;
    public final AnnotationID id;
    transient HashSet<AnnotationJoint> joints;

    public Annotation(AnnotationID id)
    {
        this.id=id;
        joints = new HashSet<>();
    }

    public void addAnnotationJoint(AnnotationJoint joint){
        joint.addAnnotation(this);
    }

    public void removeAnnotationJoint(AnnotationJoint joint){
        joint.removeAnnotation(this);
    }

    @Override
    public boolean equals(@Nullable Object obj) {
        if(obj instanceof Annotation)
        {
            Annotation a = (Annotation) obj;
            return a.id.equals(id);
        }else if (obj instanceof AnnotationID){
            AnnotationID aid = (AnnotationID) obj;
            return aid.equals(id);
        }
        return super.equals(obj);
    }

    public Set<AnnotationJoint> getAnnotationJoints(){
        return Collections.unmodifiableSet(joints);
    }

    @Override
    public int hashCode() {
        return id.hashCode();
    }

    public boolean hasJoint(AnnotationJoint aj) {
        return joints.contains(aj);
    }
}
